### PR TITLE
Create a Single Choice Question Endpoint #11

### PR DESCRIPTION
**Pull request by jdisser**
This code is complete but not tested. The database script is also not tested. The code will not yet compile due to the mismatch to the previous database schema.

### CODE REVIEW

File: src/main/java/com/haxwell/apps/quizki/dtos/CreateChoiceDTO.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	
+
+	@NotBlank
+	private String text;
+	
+
+	private boolean isCorrect;
+	
+	public CreateChoiceDTO(String text, boolean isCorrect) {
+		super();
```
Comment (haxwell) [inline src/main/java/com/haxwell/apps/quizki/dtos/CreateChoiceDTO.java:24]: The call to super() is not needed.. This class does not inherit from another (other than Object) so there is no parent constructor to call.

Comment (haxwell) [reply src/main/java/com/haxwell/apps/quizki/dtos/CreateChoiceDTO.java:24]: ..
I see you did comment out the call in other places.. you can go ahead and remove it.

File: src/main/java/com/haxwell/apps/quizki/entities/Reference.java [hunk 2]
```diff
@@ -32,7 +38,20 @@ public void setText(String str) {
     	this.text = str;
     }
     
-    public String toString() {
+    @ManyToMany(fetch = FetchType.LAZY,
```
Comment (haxwell) [inline src/main/java/com/haxwell/apps/quizki/entities/Reference.java:44]: Is there a reason this ManyToMany is defined differently than, say, the one in Choice?

File: src/main/java/com/haxwell/apps/quizki/services/QuestionServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	
+	@Autowired
+	private ReferenceRepository refRepo;
+		
+	@Autowired
+	private UserRepository userRepo;
+
+	public QuestionServiceImpl(QuestionRepository questionRepo, TopicRepository topicRepo, ReferenceRepository refRepo,
+			 UserRepository userRepo) {
+		super();
```
Comment (haxwell) [inline src/main/java/com/haxwell/apps/quizki/services/QuestionServiceImpl.java:59]: No super() necessary.

File: src/main/java/com/haxwell/apps/quizki/services/QuestionServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		for(CreateChoiceDTO ccDTO : this.choiceDTOs) {
+			Choice choice = new Choice();
+			choice.setSequence(0);		//default for Single & Multiple types need logic here for other question types
+			choice.setCorrect(ccDTO.getCorrect());
+			if(choice.getCorrect() == true)
+				++isCorrects;
+			choice.setText(ccDTO.getText());
+			this.question.getChoices().add(choice);
+		
+			System.out.println("Choice.text: " + choice.getText());
```
Comment (haxwell) [inline src/main/java/com/haxwell/apps/quizki/services/QuestionServiceImpl.java]: Be sure to remove these... System.out.printlns.

**Comment by haxwell** [comment]
Just did a quick review... added some code comments.. It looks good.. I still need to run it.

File: src/main/java/com/haxwell/apps/quizki/entities/Reference.java (hunk 2 above)
Comment (jdisser) [reply src/main/java/com/haxwell/apps/quizki/entities/Reference.java:44]: Yes, References are Many to Many so they have a Set of Questions. This way given a Reference you could get a set of all the Questions that use it. Same for Topics which also now have a set of Questions. Choices are only associated with a single Question not a set of them. Choices use a Join Column to hold the Foreign Key to the individual Question it's connected to, but that doesn't work when there are multiple Questions like here. So there is a Join Table used instead and that's the difference.

File: src/main/java/com/haxwell/apps/quizki/services/QuestionServiceImpl.java (hunk 4 above)
Comment (jdisser) [reply src/main/java/com/haxwell/apps/quizki/services/QuestionServiceImpl.java]: Yup, just for debugging.

**Comment by jdisser** [comment]
It compiles and "runs" but it's still not working yet. Just wanted to get it posted since there's a lot in it. Working on getting a boolean type to work now. Hope to have something running and in better shape by tonight. I'll comment here and on slack when it's really running.

**Comment by jdisser** [comment]
@haxwell This is NOT ready for review, but a lot closer. Working on an infinite recursion problem in the JSON generation. It's creating the questions and child entities now, but not creating the Response correctly. Have a plan for how to deal with this. Wanted to get this posted for backup. Computer not happy lately with heavy memory loads.

**Comment by jdisser** [comment]
This code is running and passing all new postman tests, cleanup remaining to be done

**Comment by jdisser** [comment]
@haxwell The code is ready for review. Postman tests were updated too.

File: src/main/java/com/haxwell/apps/quizki/entities/Reference.java (hunk 2 above)
Comment (haxwell) [reply src/main/java/com/haxwell/apps/quizki/entities/Reference.java:44]: Thank you.. and now that I go back and look, it is apparent, I mis-looked when I wrote this comment.. There actually is not a ManyToMany in Choice; it is a ManyToOne (as it should be).. and the other ManyToMany's are written this same way.. My bad, it was early.

**Comment by haxwell** [comment]
Looks great! Thanks!

**Pull request merged**